// promptfoo/dist/src/quiverai-Cj-PUa3p.js
#!/usr/bin/env node
import "./chunk-DHDDz29n.js";
import "./logger-Bzi5o47S.js";
import "./fetch-Mb7SbeM-.js";
import "./cloud-ByJVRSQI.js";
import "./types-CgrxBFgm.js";
import "./cache-DHm8Pc8l.js";
import "./util-M2MoNCUR.js";
import "./esm-Lh6czAFT.js";
import "./pythonUtils-B9JA-gsC.js";
import "./genaiTracer-foKtbVa2.js";
import { t as OpenAiChatCompletionProvider } from "./chat-CVOyOBYk.js";
import "./transform-Bn-lgBE2.js";
import "./openai-Cv9pEKxp.js";
import "./util-DHBpsbZE.js";

//#region src/providers/quiverai.ts
const QUIVERAI_API_BASE_URL = "https://api.quiver.ai/v1";
/**
* QuiverAI chat provider extends OpenAI chat completion provider.
* QuiverAI's chat API is OpenAI-compatible and excels at SVG generation.
*/
var QuiverAiChatProvider = class extends OpenAiChatCompletionProvider {
	constructor(modelName, providerOptions = {}) {
		const config = providerOptions.config || {};
		super(modelName, {
			...providerOptions,
			config: {
				...config,
				apiBaseUrl: config.apiBaseUrl || QUIVERAI_API_BASE_URL,
				apiKeyEnvar: "QUIVERAI_API_KEY"
			}
		});
	}
	id() {
		return `quiverai:${this.modelName}`;
	}
	toString() {
		return `[QuiverAI Provider ${this.modelName}]`;
	}
	getApiUrlDefault() {
		return QUIVERAI_API_BASE_URL;
	}
};
function createQuiverAiProvider(providerPath, providerOptions = {}, env) {
	const splits = providerPath.split(":");
	const modelType = splits[1];
	const effectiveEnv = env || providerOptions.env;
	return new QuiverAiChatProvider((modelType === "chat" ? splits.slice(2) : splits.slice(1)).join(":") || "arrow-0.5", {
		...providerOptions,
		env: effectiveEnv
	});
}

//#endregion
export { createQuiverAiProvider };
